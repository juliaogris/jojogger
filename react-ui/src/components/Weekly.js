import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import Row from './Jogs/Row'
import Calendar from './Jogs/Calendar'

const Weekly = ({ startDate, endDate, jogs, onDatesChange, loading }) => {
  if (loading) {
    return <h1>Loading</h1>
  }
  if (jogs.length === 0) {
    return <p className='info'>No jogs tracked.</p>
  }
  const weekly = summarizeWeeklyJogs(jogs)
  const weeklyKeys = Object.keys(weekly).sort().reverse()
  return (
    <div className='page'>
      <Calendar
        startDate={startDate}
        endDate={endDate}
        onDatesChange={onDatesChange}
      />
      <table>
        <thead>
          <tr>
            <td>Week</td>
            <td>Duration</td>
            <td>Distance</td>
            <td>Speed</td>
          </tr>
        </thead>
        <tbody>
          { weeklyKeys.map((weeklyKey) => <Row jog={weekly[weeklyKey]} key={weeklyKey} />) }
        </tbody>
      </table>
    </div>
  )
}

Weekly.propTypes = {
  jogs: PropTypes.array.isRequired
}

export default Weekly

function summarizeWeeklyJogs (jogs) {
  if (jogs.length === 0) {
    return {}
  }
  const result = {}
  const lastWeek = moment(jogs[0].date).startOf('isoWeek')
  const week = moment(jogs[jogs.length - 1].date).startOf('isoWeek')
  while (week.isSameOrBefore(lastWeek, 'isoWeek')) {
    const date = week.format('YYYY-MM-DD')
    result[date] = { date, distance: 0, duration: moment.duration(0) }
    week.add(1, 'week')
  }

  for (let i = 0; i < jogs.length; i++) {
    const jog = jogs[i]
    const date = moment(jog.date).startOf('isoWeek').format('YYYY-MM-DD')
    result[date].distance += jog.distance
    result[date].duration.add(moment.duration(jog.duration))
  }
  Object.keys(result).forEach((key) => {
    const weekly = result[key]
    let duration = String(Math.floor(weekly.duration.asHours()))
    duration += moment.utc(weekly.duration.asMilliseconds()).format(':mm:ss')
    weekly.duration = duration
  })

  return result
}

Weekly.propTypes = {
  onDatesChange: PropTypes.func.isRequired,
  jogs: PropTypes.array.isRequired
}
