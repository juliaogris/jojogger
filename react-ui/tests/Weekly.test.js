/* global it expect jest */
import React from 'react'
import ReactDOM from 'react-dom'
import Weekly from '../src/components/Weekly'
import renderer from 'react-test-renderer'

const sampleJogs = [
  { 'date': '2017-04-09', 'duration': '01:00:00', 'distance': 12.13, 'id': 'id100' },
  { 'date': '2017-04-08', 'duration': '01:00:00', 'distance': 11.5, 'id': 'id200' },
  { 'date': '2017-04-07', 'duration': '01:00:04', 'distance': 12, 'id': 'id1' },
  { 'date': '2017-04-06', 'duration': '01:00:00', 'distance': 10, 'id': 'id2' },
  { 'date': '2017-04-06', 'duration': '01:00:00', 'distance': 10, 'id': 'id3' },
  { 'date': '2017-04-06', 'duration': '01:00:00', 'distance': 10, 'id': 'id4' },
  { 'date': '2017-04-05', 'duration': '00:20:01', 'distance': 3.1, 'id': 'id5' },
  { 'date': '2017-04-05', 'duration': '01:32:00', 'distance': 6, 'id': 'id6' },
  { 'date': '2017-04-03', 'duration': '00:10:00', 'distance': 6, 'id': 'id7' },
  { 'date': '2017-04-02', 'duration': '01:00:00', 'distance': 10, 'id': 'id8' },
  { 'date': '2017-03-31', 'duration': '01:00:00', 'distance': 10, 'id': 'id9' }]

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Weekly
    onDatesChange={() => {}}
    jogs={[]}
    />, div)
})

it('renders "Weekly" and matches snapshot for no jogs', () => {
  const tree = renderer.create(<Weekly
    onDatesChange={() => {}}
    jogs={[]}
    />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders "Weekly" and matches snapshot for sample jogs', () => {
  const tree = renderer.create(<Weekly
    onDatesChange={() => {}}
    jogs={sampleJogs}
    />).toJSON()
  expect(tree).toMatchSnapshot()
})
