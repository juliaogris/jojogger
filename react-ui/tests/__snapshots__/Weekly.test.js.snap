exports[`test renders "Weekly" and matches snapshot for no jogs 1`] = `
<p
  className="info">
  No jogs tracked.
</p>
`;

exports[`test renders "Weekly" and matches snapshot for sample jogs 1`] = `
<div
  className="page">
  <div
    className="DateRangePicker">
    <div>
      <div
        className="DateRangePickerInput">
        <button
          aria-label="Focus on start date"
          className="DateRangePickerInput__calendar-icon"
          onClick={[Function]}
          type="button">
          <svg
            viewBox="0 0 1393.1 1500"
            xmlns="http://www.w3.org/2000/svg">
            <path
              d="M107.2 1392.9h241.1v-241.1H107.2v241.1zm294.7 0h267.9v-241.1H401.9v241.1zm-294.7-294.7h241.1V830.4H107.2v267.8zm294.7 0h267.9V830.4H401.9v267.8zM107.2 776.8h241.1V535.7H107.2v241.1zm616.2 616.1h267.9v-241.1H723.4v241.1zM401.9 776.8h267.9V535.7H401.9v241.1zm642.9 616.1H1286v-241.1h-241.1v241.1zm-321.4-294.7h267.9V830.4H723.4v267.8zM428.7 375V133.9c0-7.3-2.7-13.5-8-18.8-5.3-5.3-11.6-8-18.8-8h-53.6c-7.3 0-13.5 2.7-18.8 8-5.3 5.3-8 11.6-8 18.8V375c0 7.3 2.7 13.5 8 18.8 5.3 5.3 11.6 8 18.8 8h53.6c7.3 0 13.5-2.7 18.8-8 5.3-5.3 8-11.5 8-18.8zm616.1 723.2H1286V830.4h-241.1v267.8zM723.4 776.8h267.9V535.7H723.4v241.1zm321.4 0H1286V535.7h-241.1v241.1zm26.8-401.8V133.9c0-7.3-2.7-13.5-8-18.8-5.3-5.3-11.6-8-18.8-8h-53.6c-7.3 0-13.5 2.7-18.8 8-5.3 5.3-8 11.6-8 18.8V375c0 7.3 2.7 13.5 8 18.8 5.3 5.3 11.6 8 18.8 8h53.6c7.3 0 13.5-2.7 18.8-8 5.4-5.3 8-11.5 8-18.8zm321.5-53.6v1071.4c0 29-10.6 54.1-31.8 75.3-21.2 21.2-46.3 31.8-75.3 31.8H107.2c-29 0-54.1-10.6-75.3-31.8C10.6 1447 0 1421.9 0 1392.9V321.4c0-29 10.6-54.1 31.8-75.3s46.3-31.8 75.3-31.8h107.2v-80.4c0-36.8 13.1-68.4 39.3-94.6S311.4 0 348.3 0h53.6c36.8 0 68.4 13.1 94.6 39.3 26.2 26.2 39.3 57.8 39.3 94.6v80.4h321.5v-80.4c0-36.8 13.1-68.4 39.3-94.6C922.9 13.1 954.4 0 991.3 0h53.6c36.8 0 68.4 13.1 94.6 39.3s39.3 57.8 39.3 94.6v80.4H1286c29 0 54.1 10.6 75.3 31.8 21.2 21.2 31.8 46.3 31.8 75.3z" />
          </svg>
        </button>
        <div
          className="DateInput">
          <input
            aria-describedby=""
            aria-label="Start Date"
            autoComplete="off"
            className="DateInput__input needsclick"
            disabled={false}
            id="startDate"
            name="startDate"
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Start Date"
            readOnly={false}
            required={false}
            type="text"
            value="" />
          
          <div
            className="DateInput__display-text">
            Start Date
          </div>
        </div>
        <div
          className="DateRangePickerInput__arrow">
          <span
            className="drp">
            →
          </span>
        </div>
        <div
          className="DateInput">
          <input
            aria-describedby=""
            aria-label="End Date"
            autoComplete="off"
            className="DateInput__input needsclick"
            disabled={false}
            id="endDate"
            name="endDate"
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="End Date"
            readOnly={false}
            required={false}
            type="text"
            value="" />
          
          <div
            className="DateInput__display-text">
            End Date
          </div>
        </div>
        <button
          className="DateRangePickerInput__clear-dates DateRangePickerInput__clear-dates--hide"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          type="button">
          <span
            className="screen-reader-only">
            Clear Dates
          </span>
          <div
            className="DateRangePickerInput__close-icon">
            <span
              className="drp">
              ⨉
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
  <table>
    <thead>
      <tr>
        <td>
          Week
        </td>
        <td>
          Duration
        </td>
        <td>
          Distance
        </td>
        <td>
          Speed
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          2017-04-03
        </td>
        <td>
          8:02:05
        </td>
        <td>
          80.73
        </td>
        <td>
          10.05
        </td>
      </tr>
      <tr>
        <td>
          2017-03-27
        </td>
        <td>
          2:00:00
        </td>
        <td>
          20.00
        </td>
        <td>
          10.00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
